infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/StreamsSnapshotReader.java

```diff
@@ -56,68 +72,115 @@
      * @param stream
-     * @param entries
+     * @param entryList
      * @return
      */
-    private LogReplicationEntry generateMessage(OpaqueStreamIterator stream, List<SMREntry> entries, UUID snapshotRequestId) {
+    private LogReplicationEntry generateMessage(OpaqueStreamIterator stream, SMREntryList entryList, UUID snapshotRequestId) {
         currentMsgTs = stream.maxVersion;
-        OpaqueEntry opaqueEntry = generateOpaqueEntry(currentMsgTs, stream.uuid, entries);
+        OpaqueEntry opaqueEntry = generateOpaqueEntry(currentMsgTs, stream.uuid, entryList);
         if (!stream.iterator.hasNext()) {
             //mark the end of the current stream.
             currentMsgTs = snapshotTimestamp;
         }
 
-        ByteBuf buf = Unpooled.buffer();
+        ByteBuf buf = Unpooled.buffer(entryList.size);
         OpaqueEntry.serialize(buf, opaqueEntry);
 
-        org.corfudb.protocols.wireprotocol.logreplication.LogReplicationEntry txMsg = new org.corfudb.protocols.wireprotocol.logreplication.LogReplicationEntry
-                (MessageType.SNAPSHOT_MESSAGE, topologyConfigId, snapshotRequestId, currentMsgTs,
-                preMsgTs, snapshotTimestamp, sequence, buf.array());
+        // Make a new buf backed by data only.
+        ByteBuf newBuf = Unpooled.wrappedBuffer(buf.array(), 0, buf.writerIndex());
+
+        LogReplicationEntry txMsg = new LogReplicationEntry(MessageType.SNAPSHOT_MESSAGE, topologyConfigId, snapshotRequestId, currentMsgTs,
+                preMsgTs, snapshotTimestamp, sequence, newBuf.array());
+
         preMsgTs = currentMsgTs;
         sequence++;
-        log.debug("Generate TxMsg {}", txMsg.getMetadata());
+
+        log.trace("txMsg {} deepsize size {} buf size {} buf.writerIndex {} entryList.size {} numEntries {} deepSize size {}",
+                txMsg.getMetadata(), MetricsUtils.sizeOf.deepSizeOf(txMsg), MetricsUtils.sizeOf.deepSizeOf(buf),
+                buf.writerIndex(), entryList.getSize(), entryList.getSmrEntries().size(), MetricsUtils.sizeOf.deepSizeOf(entryList.smrEntries));
+
         return txMsg;
     }
 
     /**
-     * Read numEntries from the current stream.
+     * Given a list of SMREntries, calculate the total size.
+     * @param smrEntries
+     * @return
+     */
+    private int calculateSize(List<SMREntry> smrEntries) {
+        int size = 0;
+        for (SMREntry entry : smrEntries) {
+            size += entry.getSerializedSize();
+        }
+
+        log.trace("current entry size {}", size);
+        return size;
+    }
+
+    /**
+     * Read log data from the current stream until the sum of all SMR entries's size reaches the maxDataSize.
      * @param stream
-     * @param numEntries
      * @return
      */
-    private List<SMREntry> next(OpaqueStreamIterator stream, int numEntries) {
-        //if it is the end of the stream, set an end of stream mark, the current
+    private SMREntryList next(OpaqueStreamIterator stream) {
```

Comment:
Can we add a unit test on this? as it is core to snapshot sending protocol.

Reply:
You can create a PR track that.